### PR TITLE
Supporting use-case 1.xi (a-c): a model with variables

### PR DESCRIPTION
**Pull request by dladd**
This request (upon completion) implements tests and code support for use-case 1.xi: a model with variables

Addresses #63

### CODE REVIEW

File: tests/variable/variable.cpp [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+limitations under the License.Some license of other
+*/
+
+#include "gtest/gtest.h"
+
+#include <libcellml>
+
+
+TEST(Variable, serialise) {
+    const std::string e = "";
```
Comment (nickerso) [inline tests/variable/variable.cpp]: as with the other classes, I would expect a new variable to produce at least an empty `<variable/>` element in the XML serialisation.

File: tests/variable/variable.cpp [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    u->setName("dimensionless");
+    v->setUnits(u);
+
+    std::string a = m.serialise(libcellml::FORMAT_XML);
+    EXPECT_EQ(e, a);
+    EXPECT_EQ("valid_name", v.getName());
+}
+
+TEST(Variable, invalidVariableName) {
+    const std::string in = "invalid_name";
```
Comment (nickerso) [inline tests/variable/variable.cpp]: I should go back and check the component test, but this should actually be an invalid name like `invalid name` :)

File: tests/variable/variable.cpp [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    libcellml::ComponentPtr c = std::make_shared<libcellml::Component>();
+    c->setName(in);
+    m.addComponent(c);
+
+    libcellml::VariablePtr v = std::make_shared<libcellml::Variable>();
+    v->setName(in);
+    c->addVariable(v);
+
+    libcellml::UnitsPtr u = std::make_shared<libcellml::Units>();
+    u->setName("invalid_name");
```
Comment (nickerso) [inline tests/variable/variable.cpp]: again, should be an invalid name

**Comment by nickerso** [comment]
good start. Is the buildbot testing not working currently? shouldn't there be a green tick or red cross on the current state of the pull request? (@hsorby)

**Comment by hsorby** [comment]
It would appear that Buildbot has given up @nickerso I shall investigate.

**Comment by dladd** [comment]
This now passes all tests and adds the features as set out in #63. 

However, there were two applications I implemented aspects of as part of the outlined object model (see #61) that were overlooked in the use-cases (see #63). These were (1) components having multiple variables and (2) variables having initial values.

Adding tests and support for these should be relatively straightforward. I propose:
- **1.xi.a.1** a model from 1.xi.a with a valid variable initial value of 0.0
- **1.xi.d** a model with a single component containing two variables named "variable_a" and "variable_b" with units dimensionless.
  - **1.xi.d.1** a model from 1.xi.d with variable initial values of -1.0 and 1.0, respectively.

The use-case currently listed as 1.xi.d dealing with interface connections could be shifted into 1.xii as discussed in (#63), with the rest of the use-cases to be shifted accordingly

**Comment by hsorby** [comment]
I see some future code is added here (which consequently drops the coverage, http://autotest.bioeng.auckland.ac.nz/libcellml-buildbot/d0a57450d6e63d5e40ce534cd7409d254fde7d9b/dox/coverage/).  Should be removed for now and added back in at the appropriate time.

File: src/variable.cpp [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+std::string Variable::doSerialisation(FORMATS format) const
+{
+    std::string repr = "";
+    if (format == FORMAT_XML) {
+        repr += "<variable";
+        if (getName().length()) {
+            repr += " name=\"" + getName() + "\"";
+            if (getUnits() != nullptr) {
+                repr += " units=\"" + getUnits()->getName() + "\"";
+            }
```
Comment (hsorby) [inline src/variable.cpp]: This should be independent of whether that variable has a name or not.

**Comment by dladd** [comment]
Thanks, @hsorby. We should now be back up to 100% coverage on this. 

Tests now also cover multiple variables in a component. That leaves variable initial values to be tacked on to this, unless others think that is better left for a separate pull.

File: src/variable.cpp (hunk 4 above)
Comment (jonc125) [inline src/variable.cpp]: And hence a test added where a variable has units but no name.
(When we come to initial values, we should test more combinations.)

**Comment by jonc125** [comment]
I'd put it in a separate pull given the title of this one :)

Haven't checked thoroughly, but apart from the extra test I mentioned this seems sane.

**Comment by dladd** [comment]
Thanks, @jonc125. I've now added the units-only test and will go with a separate pull for initial values (see #67).

**Pull request merged**